### PR TITLE
[Composed Recipes] IC integration

### PR DESCRIPTION
**Pull request by dbogunowicz**
Related to Asana task: https://app.asana.com/0/0/1202094848443166/f

Tested manually:

1. using command 
```
sparseml/pytorch/image_classification/train.py
--recipe-path
/Users/damian/Code/nm/sparseml/integrations/pytorch/recipes/resnet50-imagenette-pruned.md
--arch-key
resnet50
--dataset
imagenette
--dataset-path
data
--train-batch-size
128
--test-batch-size
128
--save-epochs
2
```

trained for 2 epochs to generate a checkpoint file. Checkpoint file's state contains recipe:

```
version: 1.1.0

__metadata__:
  arch_key: resnet50
  dataset: imagenette
  device: cuda:0,1,2,3
  pretrained: True
  test_batch_size: 128
  train_batch_size: 128
  framework_metadata:
    python_version: 3.8.10
    sparseml_version: 0.12.0.20220505
    torch_version: 1.9.1+cu111

modifiers:
    - !GMPruningModifier
        end_epoch: 8.0
        final_sparsity: 0.8
        global_sparsity: False
        init_sparsity: 0.05
        inter_func: cubic
        leave_enabled: True
        mask_type: unstructured
        params: ['sections.0.0.conv1.weight', 'sections.0.0.conv2.weight', 'sections.0.0.conv3.weight', 'sections.0.0.identity.conv.weight', 'sections.0.1.conv1.weight', 'sections.0.1.conv3.weight', 'sections.0.2.conv1.weight', 'sections.0.2.conv3.weight', 'sections.1.0.conv1.weight', 'sections.1.0.conv3.weight', 'sections.1.2.conv3.weight', 'sections.1.3.conv1.weight', 'sections.2.0.conv1.weight', 'sections.3.0.conv1.weight', 'classifier.fc.weight']
        start_epoch: 1.0
        update_frequency: 0.5

    - !GMPruningModifier
        end_epoch: 8.0
        final_sparsity: 0.9
        global_sparsity: False
        init_sparsity: 0.05
        inter_func: cubic
        leave_enabled: True
        mask_type: unstructured
        params: ['sections.1.0.conv2.weight', 'sections.2.0.conv2.weight', 'sections.2.1.conv2.weight', 'sections.2.2.conv2.weight', 'sections.2.3.conv2.weight', 'sections.2.4.conv2.weight', 'sections.2.5.conv2.weight', 'sections.3.0.conv2.weight', 'sections.3.0.conv3.weight', 'sections.3.0.identity.conv.weight', 'sections.3.1.conv2.weight', 'sections.3.1.conv3.weight', 'sections.3.2.conv2.weight', 'sections.3.2.conv3.weight']
        start_epoch: 1.0
        update_frequency: 0.5

    - !GMPruningModifier
        end_epoch: 8.0
        final_sparsity: 0.85
        global_sparsity: False
        init_sparsity: 0.05
        inter_func: cubic
        leave_enabled: True
        mask_type: unstructured
        params: ['sections.0.1.conv2.weight', 'sections.0.2.conv2.weight', 'sections.1.0.identity.conv.weight', 'sections.1.1.conv1.weight', 'sections.1.1.conv2.weight', 'sections.1.1.conv3.weight', 'sections.1.2.conv1.weight', 'sections.1.2.conv2.weight', 'sections.1.3.conv2.weight', 'sections.1.3.conv3.weight', 'sections.2.0.conv3.weight', 'sections.2.0.identity.conv.weight', 'sections.2.1.conv1.weight', 'sections.2.1.conv3.weight', 'sections.2.2.conv1.weight', 'sections.2.2.conv3.weight', 'sections.2.3.conv1.weight', 'sections.2.3.conv3.weight', 'sections.2.4.conv1.weight', 'sections.2.4.conv3.weight', 'sections.2.5.conv1.weight', 'sections.2.5.conv3.weight', 'sections.3.1.conv1.weight', 'sections.3.2.conv1.weight']
        start_epoch: 1.0
        update_frequency: 0.5

    - !EpochRangeModifier
        end_epoch: 10
        start_epoch: 0.0

    - !SetLearningRateModifier
        constant_logging: False
        end_epoch: -1.0
        learning_rate: 0.008
        start_epoch: 0.0
```

2. using command 
```
sparseml/pytorch/image_classification/train.py
--recipe-path
/Users/damian/Code/nm/sparseml/integrations/pytorch/recipes/resnet50-imagenette-pruned.md
--arch-key
resnet50
--dataset
imagenette
--dataset-path
data
--train-batch-size
128
--test-batch-size
128
--save-epochs
2
--checkpoint-path
/home/damian/sparseml/pytorch_vision/resnet50_imagenette__09/framework/checkpoint-best.pth
```

trained for 2 epochs to generate a checkpoint file. Checkpoint file's state contains recipe:

```
version: 1.1.0

stage_0:
  __metadata__:
    arch_key: resnet50
    dataset: imagenette
    device: cuda:0,1,2,3
    pretrained: True
    test_batch_size: 128
    train_batch_size: 128
    framework_metadata:
      python_version: 3.8.10
      sparseml_version: 0.12.0.20220505
      torch_version: 1.9.1+cu111

  stage_0_modifiers:
      - !GMPruningModifier
          end_epoch: 8.0
          final_sparsity: 0.8
          global_sparsity: False
          init_sparsity: 0.05
          inter_func: cubic
          leave_enabled: True
          mask_type: unstructured
          params: ['sections.0.0.conv1.weight', 'sections.0.0.conv2.weight', 'sections.0.0.conv3.weight', 'sections.0.0.identity.conv.weight', 'sections.0.1.conv1.weight', 'sections.0.1.conv3.weight', 'sections.0.2.conv1.weight', 'sections.0.2.conv3.weight', 'sections.1.0.conv1.weight', 'sections.1.0.conv3.weight', 'sections.1.2.conv3.weight', 'sections.1.3.conv1.weight', 'sections.2.0.conv1.weight', 'sections.3.0.conv1.weight', 'classifier.fc.weight']
          start_epoch: 1.0
          update_frequency: 0.5
  
      - !GMPruningModifier
          end_epoch: 8.0
          final_sparsity: 0.9
          global_sparsity: False
          init_sparsity: 0.05
          inter_func: cubic
          leave_enabled: True
          mask_type: unstructured
          params: ['sections.1.0.conv2.weight', 'sections.2.0.conv2.weight', 'sections.2.1.conv2.weight', 'sections.2.2.conv2.weight', 'sections.2.3.conv2.weight', 'sections.2.4.conv2.weight', 'sections.2.5.conv2.weight', 'sections.3.0.conv2.weight', 'sections.3.0.conv3.weight', 'sections.3.0.identity.conv.weight', 'sections.3.1.conv2.weight', 'sections.3.1.conv3.weight', 'sections.3.2.conv2.weight', 'sections.3.2.conv3.weight']
          start_epoch: 1.0
          update_frequency: 0.5
  
      - !GMPruningModifier
          end_epoch: 8.0
          final_sparsity: 0.85
          global_sparsity: False
          init_sparsity: 0.05
          inter_func: cubic
          leave_enabled: True
          mask_type: unstructured
          params: ['sections.0.1.conv2.weight', 'sections.0.2.conv2.weight', 'sections.1.0.identity.conv.weight', 'sections.1.1.conv1.weight', 'sections.1.1.conv2.weight', 'sections.1.1.conv3.weight', 'sections.1.2.conv1.weight', 'sections.1.2.conv2.weight', 'sections.1.3.conv2.weight', 'sections.1.3.conv3.weight', 'sections.2.0.conv3.weight', 'sections.2.0.identity.conv.weight', 'sections.2.1.conv1.weight', 'sections.2.1.conv3.weight', 'sections.2.2.conv1.weight', 'sections.2.2.conv3.weight', 'sections.2.3.conv1.weight', 'sections.2.3.conv3.weight', 'sections.2.4.conv1.weight', 'sections.2.4.conv3.weight', 'sections.2.5.conv1.weight', 'sections.2.5.conv3.weight', 'sections.3.1.conv1.weight', 'sections.3.2.conv1.weight']
          start_epoch: 1.0
          update_frequency: 0.5
  
      - !EpochRangeModifier
          end_epoch: 10
          start_epoch: 0.0
  
      - !SetLearningRateModifier
          constant_logging: False
          end_epoch: 10
          learning_rate: 0.008
          start_epoch: 0.0
  

stage_1:
  __metadata__:
    arch_key: resnet50
    dataset: imagenette
    device: cuda:0,1,2,3
    pretrained: True
    test_batch_size: 128
    train_batch_size: 128
    framework_metadata:
      python_version: 3.8.10
      sparseml_version: 0.12.0.20220505
      torch_version: 1.9.1+cu111

  stage_1_modifiers:
      - !GMPruningModifier
          end_epoch: 18.0
          final_sparsity: 0.9
          global_sparsity: False
          init_sparsity: 0.05
          inter_func: cubic
          leave_enabled: True
          mask_type: unstructured
          params: ['sections.1.0.conv2.weight', 'sections.2.0.conv2.weight', 'sections.2.1.conv2.weight', 'sections.2.2.conv2.weight', 'sections.2.3.conv2.weight', 'sections.2.4.conv2.weight', 'sections.2.5.conv2.weight', 'sections.3.0.conv2.weight', 'sections.3.0.conv3.weight', 'sections.3.0.identity.conv.weight', 'sections.3.1.conv2.weight', 'sections.3.1.conv3.weight', 'sections.3.2.conv2.weight', 'sections.3.2.conv3.weight']
          start_epoch: 11.0
          update_frequency: 0.5
  
      - !GMPruningModifier
          end_epoch: 18.0
          final_sparsity: 0.85
          global_sparsity: False
          init_sparsity: 0.05
          inter_func: cubic
          leave_enabled: True
          mask_type: unstructured
          params: ['sections.0.1.conv2.weight', 'sections.0.2.conv2.weight', 'sections.1.0.identity.conv.weight', 'sections.1.1.conv1.weight', 'sections.1.1.conv2.weight', 'sections.1.1.conv3.weight', 'sections.1.2.conv1.weight', 'sections.1.2.conv2.weight', 'sections.1.3.conv2.weight', 'sections.1.3.conv3.weight', 'sections.2.0.conv3.weight', 'sections.2.0.identity.conv.weight', 'sections.2.1.conv1.weight', 'sections.2.1.conv3.weight', 'sections.2.2.conv1.weight', 'sections.2.2.conv3.weight', 'sections.2.3.conv1.weight', 'sections.2.3.conv3.weight', 'sections.2.4.conv1.weight', 'sections.2.4.conv3.weight', 'sections.2.5.conv1.weight', 'sections.2.5.conv3.weight', 'sections.3.1.conv1.weight', 'sections.3.2.conv1.weight']
          start_epoch: 11.0
          update_frequency: 0.5
  
      - !GMPruningModifier
          end_epoch: 18.0
          final_sparsity: 0.8
          global_sparsity: False
          init_sparsity: 0.05
          inter_func: cubic
          leave_enabled: True
          mask_type: unstructured
          params: ['sections.0.0.conv1.weight', 'sections.0.0.conv2.weight', 'sections.0.0.conv3.weight', 'sections.0.0.identity.conv.weight', 'sections.0.1.conv1.weight', 'sections.0.1.conv3.weight', 'sections.0.2.conv1.weight', 'sections.0.2.conv3.weight', 'sections.1.0.conv1.weight', 'sections.1.0.conv3.weight', 'sections.1.2.conv3.weight', 'sections.1.3.conv1.weight', 'sections.2.0.conv1.weight', 'sections.3.0.conv1.weight', 'classifier.fc.weight']
          start_epoch: 11.0
          update_frequency: 0.5
  
      - !EpochRangeModifier
          end_epoch: 20
          start_epoch: 10.0
  
      - !SetLearningRateModifier
          constant_logging: False
          end_epoch: -1.0
          learning_rate: 0.008
          start_epoch: 10.0
 ```

### CODE REVIEW

File: src/sparseml/pytorch/image_classification/utils/trainer.py [hunk 1]
```diff
@@ -151,6 +159,22 @@ def __init__(
...
+        required_keys = ["recipe", "state_dict"]
+        for key in required_keys:
+            if key not in checkpoint_state.keys():
+                raise ValueError(
+                    f"The expected key {key} was not found "
+                    f"in the checkpoint from {self.checkpoint_path}"
+                )
+        checkpoint_recipe = checkpoint_state["recipe"]
+        checkpoint_manager = ScheduledModifierManager.from_yaml(checkpoint_recipe)
+        checkpoint_manager.apply(self.model)
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/image_classification/utils/trainer.py:173]: @rahul-tuli how is it different from apply structure.

Comment (dbogunowicz) [inline src/sparseml/pytorch/image_classification/utils/trainer.py:173]: @rahul-tuli how is it different from apply structure.

**Review by KSGulin — DISMISSED** [review]
Great stuff!

**Review by bfineran — CHANGES_REQUESTED** [review]
@dbogunowicz LGTM, looks like we may be missing one more piece - need to call `apply_structure` on the checkpoint recipe before loading weights
(see #481 for an early hack of this).

The affected function is `load_model` in `pytorch/utils/model.py`. Can close #481 once that functionality is added here

**Comment by dbogunowicz** [comment]
> @dbogunowicz LGTM, looks like we may be missing one more piece - need to call `apply_structure` on the checkpoint recipe before loading weights (see #481 for an early hack of this).
> 
> The affected function is `load_model` in `pytorch/utils/model.py`. Can close #481 once that functionality is added here

Thank you for reminding me. This was buried in my git stash and committed...
While at it, I spent some time resolving merge conflicts and integrating with click. The current version is manually tested and ready for next round of reviews.

File: src/sparseml/pytorch/utils/model.py [hunk 2]
```diff
@@ -85,6 +85,14 @@ def load_model(
             )[0]
     model_dict = torch.load(path, map_location="cpu")
     current_dict = model.state_dict()
+    recipe = model_dict.get("recipe")
+
+    if recipe:
+        from sparseml.pytorch.optim import ScheduledModifierManager
+
+        epoch = model_dict.get("epoch", 0.0)
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/model.py:93]:
```suggestion
        epoch = model_dict.get("epoch", float("inf"))
```

**Review by bfineran — COMMENTED** [review]
lgtm pending suggestion

**Review by bfineran — APPROVED** [review]

**Review by KSGulin — APPROVED** [review]

**Pull request merged**